third_party/base32_crockford.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2025, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include "base32_crockford.h"
+
+#include <ctype.h>
+#include <string.h>
+
+/* Crockford Base32 alphabet. */
+static const char crockford_alphabet[] = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
+
+/* Fast mapping from ASCII char to 5-bit Crockford value, -1 for invalid. */
+static int8_t crockford_inv[256];
+
+__attribute__((constructor))
+static void
+base32_crockford_init(void)
+{
+	for (int i = 0; i < 256; i++)
+		crockford_inv[i] = -1;
+
+	for (int i = 0; crockford_alphabet[i]; i++) {
+		crockford_inv[(uint8_t)crockford_alphabet[i]] = i;
+		crockford_inv[(uint8_t)tolower(crockford_alphabet[i])] = i;
+	}
+
+	/* Ambiguous aliases */
+	crockford_inv['I'] = crockford_inv['i'] = 1;
+	crockford_inv['L'] = crockford_inv['l'] = 1;
+	crockford_inv['O'] = crockford_inv['o'] = 0;
+}
+
+void
+base32_crockford_encode(const uint8_t *in, size_t len, char *out)
+{
+	uint32_t acc = 0;
+	int bits = 0;
+	size_t pos = 0;
+
+	for (size_t i = 0; i < len; i++) {
+		acc = (acc << 8) | in[i];
+		bits += 8;
+
+		while (bits >= 5) {
+			bits -= 5;
+			out[pos++] = crockford_alphabet[(acc >> bits) & 31];
+		}
+	}
+
+	if (bits > 0)
+		out[pos++] = crockford_alphabet[(acc << (5 - bits)) & 31];
+
+	out[pos] = '\0';
+}
+
+int
+base32_crockford_decode(const char *in, uint8_t *out, size_t out_len)
+{
+	uint32_t acc = 0;
+	int bits = 0;
+	size_t pos = 0;
+
+	for (size_t i = 0; in[i] != '\0'; i++) {
+		int8_t v = crockford_inv[(uint8_t)in[i]];
+		if (v < 0)
+			return -1;
+
+		acc = (acc << 5) | v;
+		bits += 5;
+
+		if (bits >= 8) {
+			bits -= 8;
+			if (pos >= out_len)
+				return -1;
+			out[pos++] = (acc >> bits) & 0xff;
+		}
+	}
+	return 0;
```

Comment:
I guess we should report error if `acc != 0`. This is case of invalid input also.